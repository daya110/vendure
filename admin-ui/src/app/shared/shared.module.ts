import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { ClarityModule, ClrFormsNextModule } from '@clr/angular';
import { NgSelectModule } from '@ng-select/ng-select';
import { TranslateModule } from '@ngx-translate/core';
import { NgxPaginationModule } from 'ngx-pagination';

import {
    ActionBarComponent,
    ActionBarLeftComponent,
    ActionBarRightComponent,
} from './components/action-bar/action-bar.component';
import { AffixedInputComponent } from './components/affixed-input/affixed-input.component';
import { PercentageSuffixInputComponent } from './components/affixed-input/percentage-suffix-input.component';
import { ChipComponent } from './components/chip/chip.component';
import { CurrencyInputComponent } from './components/currency-input/currency-input.component';
import { CustomFieldControlComponent } from './components/custom-field-control/custom-field-control.component';
import { DataTableColumnComponent } from './components/data-table/data-table-column.component';
import { DataTableComponent } from './components/data-table/data-table.component';
import { FormFieldControlDirective } from './components/form-field/form-field-control.directive';
import { FormFieldComponent } from './components/form-field/form-field.component';
import { FormItemComponent } from './components/form-item/form-item.component';
import { ItemsPerPageControlsComponent } from './components/items-per-page-controls/items-per-page-controls.component';
import { LanguageSelectorComponent } from './components/language-selector/language-selector.component';
import { DialogButtonsDirective } from './components/modal-dialog/dialog-buttons.directive';
import { DialogComponentOutletComponent } from './components/modal-dialog/dialog-component-outlet.component';
import { DialogTitleDirective } from './components/modal-dialog/dialog-title.directive';
import { ModalDialogComponent } from './components/modal-dialog/modal-dialog.component';
import { PaginationControlsComponent } from './components/pagination-controls/pagination-controls.component';
import { SelectToggleComponent } from './components/select-toggle/select-toggle.component';
import { TableRowActionComponent } from './components/table-row-action/table-row-action.component';
import { FileSizePipe } from './pipes/file-size.pipe';
import { ModalService } from './providers/modal/modal.service';

const IMPORTS = [
    ClarityModule,
    ClrFormsNextModule,
    CommonModule,
    FormsModule,
    ReactiveFormsModule,
    RouterModule,
    NgSelectModule,
    NgxPaginationModule,
    TranslateModule,
];

const DECLARATIONS = [
    ActionBarComponent,
    ActionBarLeftComponent,
    ActionBarRightComponent,
    AffixedInputComponent,
    ChipComponent,
    CurrencyInputComponent,
    CustomFieldControlComponent,
    DataTableComponent,
    DataTableColumnComponent,
    ItemsPerPageControlsComponent,
    PaginationControlsComponent,
    TableRowActionComponent,
    FileSizePipe,
    FormFieldComponent,
    FormFieldControlDirective,
    FormItemComponent,
    ModalDialogComponent,
    PercentageSuffixInputComponent,
    DialogComponentOutletComponent,
    DialogButtonsDirective,
    DialogTitleDirective,
    SelectToggleComponent,
    LanguageSelectorComponent,
];

@NgModule({
    imports: IMPORTS,
    exports: [...IMPORTS, ...DECLARATIONS],
    declarations: DECLARATIONS,
    providers: [
        // This needs to be shared, since lazy-loaded
        // modules have their own entryComponents which
        // are unknown to the CoreModule instance of ModalService.
        // See https://github.com/angular/angular/issues/14324#issuecomment-305650763
        ModalService,
    ],
    entryComponents: [ModalDialogComponent],
})
export class SharedModule {}
