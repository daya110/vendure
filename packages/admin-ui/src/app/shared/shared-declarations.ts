export {
    ActionBarComponent,
    ActionBarLeftComponent,
    ActionBarRightComponent,
} from './components/action-bar/action-bar.component';
export { AffixedInputComponent } from './components/affixed-input/affixed-input.component';
export { AssetFileInputComponent } from './components/asset-file-input/asset-file-input.component';
export { AssetGalleryComponent } from './components/asset-gallery/asset-gallery.component';
export { AssetPickerDialogComponent } from './components/asset-picker-dialog/asset-picker-dialog.component';
export { PercentageSuffixInputComponent } from './components/affixed-input/percentage-suffix-input.component';
export {
    ChannelAssignmentControlComponent,
} from './components/channel-assignment-control/channel-assignment-control.component';
export { ChannelBadgeComponent } from './components/channel-badge/channel-badge.component';
export { ChannelLabelPipe } from './pipes/channel-label.pipe';
export { ChipComponent } from './components/chip/chip.component';
export { ConfigurableInputComponent } from './components/configurable-input/configurable-input.component';
export { CurrencyInputComponent } from './components/currency-input/currency-input.component';
export {
    CustomFieldControlComponent,
} from './components/custom-field-control/custom-field-control.component';
export { CustomerLabelComponent } from './components/customer-label/customer-label.component';
export { DataTableColumnComponent } from './components/data-table/data-table-column.component';
export { DataTableComponent } from './components/data-table/data-table.component';
export { DropdownItemDirective } from './components/dropdown/dropdown-item.directive';
export { DropdownMenuComponent } from './components/dropdown/dropdown-menu.component';
export { DropdownTriggerDirective } from './components/dropdown/dropdown-trigger.directive';
export { DropdownComponent } from './components/dropdown/dropdown.component';
export { FacetValueChipComponent } from './components/facet-value-chip/facet-value-chip.component';
export {
    FacetValueSelectorComponent,
} from './components/facet-value-selector/facet-value-selector.component';
export { FormFieldControlDirective } from './components/form-field/form-field-control.directive';
export { FormFieldComponent } from './components/form-field/form-field.component';
export { FormItemComponent } from './components/form-item/form-item.component';
export { FormattedAddressComponent } from './components/formatted-address/formatted-address.component';
export { IfDefaultChannelActiveDirective } from './directives/if-default-channel-active.directive';
export { IfMultichannelDirective } from './directives/if-multichannel.directive';
export { IfPermissionsDirective } from './directives/if-permissions.directive';
export {
    ItemsPerPageControlsComponent,
} from './components/items-per-page-controls/items-per-page-controls.component';
export { LabeledDataComponent } from './components/labeled-data/labeled-data.component';
export { LanguageSelectorComponent } from './components/language-selector/language-selector.component';
export { DialogButtonsDirective } from './components/modal-dialog/dialog-buttons.directive';
export { DialogComponentOutletComponent } from './components/modal-dialog/dialog-component-outlet.component';
export { DialogTitleDirective } from './components/modal-dialog/dialog-title.directive';
export { ExtensionHostComponent } from './components/extension-host/extension-host.component';
export * from './components/extension-host/extension-host-config';
export { ModalDialogComponent } from './components/modal-dialog/modal-dialog.component';
export { ObjectTreeComponent } from './components/object-tree/object-tree.component';
export { OrderStateLabelComponent } from './components/order-state-label/order-state-label.component';
export { PaginationControlsComponent } from './components/pagination-controls/pagination-controls.component';
export { RichTextEditorComponent } from './components/rich-text-editor/rich-text-editor.component';
export { SelectToggleComponent } from './components/select-toggle/select-toggle.component';
export { SimpleDialogComponent } from './components/simple-dialog/simple-dialog.component';
export { TableRowActionComponent } from './components/table-row-action/table-row-action.component';
export { TitleInputComponent } from './components/title-input/title-input.component';
export { EntityInfoComponent } from './components/entity-info/entity-info.component';
export { DatetimePickerComponent } from './components/datetime-picker/datetime-picker.component';
export { CurrencyNamePipe } from './pipes/currency-name.pipe';
export { FileSizePipe } from './pipes/file-size.pipe';
export { SentenceCasePipe } from './pipes/sentence-case.pipe';
export { SortPipe } from './pipes/sort.pipe';
export { StringToColorPipe } from './pipes/string-to-color.pipe';
