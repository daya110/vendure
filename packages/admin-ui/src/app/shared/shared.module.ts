import { OverlayModule } from '@angular/cdk/overlay';
import { CommonModule } from '@angular/common';
import { CUSTOM_ELEMENTS_SCHEMA, NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { ClarityModule } from '@clr/angular';
import { NgSelectModule } from '@ng-select/ng-select';
import { TranslateModule } from '@ngx-translate/core';
import { NgxPaginationModule } from 'ngx-pagination';

import { ActionBarItemsComponent } from './components/action-bar-items/action-bar-items.component';
import {
    ActionBarComponent,
    ActionBarLeftComponent,
    ActionBarRightComponent,
} from './components/action-bar/action-bar.component';
import { DisabledDirective } from './directives/disabled.directive';
import { HasPermissionPipe } from './pipes/has-permission.pipe';
import { ModalService } from './providers/modal/modal.service';
import { CanDeactivateDetailGuard } from './providers/routing/can-deactivate-detail-guard';
import {
    AffixedInputComponent,
    AssetFileInputComponent,
    AssetGalleryComponent,
    AssetPickerDialogComponent,
    ChannelAssignmentControlComponent,
    ChannelBadgeComponent,
    ChannelLabelPipe,
    ChipComponent,
    ConfigurableInputComponent,
    CurrencyInputComponent,
    CurrencyNamePipe,
    CustomerLabelComponent,
    CustomFieldControlComponent,
    CustomFieldLabelPipe,
    DataTableColumnComponent,
    DataTableComponent,
    DatetimePickerComponent,
    DialogButtonsDirective,
    DialogComponentOutletComponent,
    DialogTitleDirective,
    DropdownComponent,
    DropdownItemDirective,
    DropdownMenuComponent,
    DropdownTriggerDirective,
    EntityInfoComponent,
    ExtensionHostComponent,
    FacetValueChipComponent,
    FacetValueSelectorComponent,
    FileSizePipe,
    FormattedAddressComponent,
    FormFieldComponent,
    FormFieldControlDirective,
    FormItemComponent,
    IfDefaultChannelActiveDirective,
    IfMultichannelDirective,
    IfPermissionsDirective,
    ItemsPerPageControlsComponent,
    LabeledDataComponent,
    LanguageSelectorComponent,
    ModalDialogComponent,
    ObjectTreeComponent,
    OrderStateLabelComponent,
    PaginationControlsComponent,
    PercentageSuffixInputComponent,
    RichTextEditorComponent,
    SelectToggleComponent,
    SentenceCasePipe,
    SimpleDialogComponent,
    SortPipe,
    StringToColorPipe,
    TableRowActionComponent,
    TitleInputComponent,
} from './shared-declarations';

const IMPORTS = [
    ClarityModule,
    CommonModule,
    FormsModule,
    ReactiveFormsModule,
    RouterModule,
    NgSelectModule,
    NgxPaginationModule,
    TranslateModule,
    OverlayModule,
];

const DECLARATIONS = [
    ActionBarComponent,
    ActionBarLeftComponent,
    ActionBarRightComponent,
    ConfigurableInputComponent,
    AffixedInputComponent,
    ChipComponent,
    CurrencyInputComponent,
    CurrencyNamePipe,
    CustomerLabelComponent,
    CustomFieldControlComponent,
    DataTableComponent,
    DataTableColumnComponent,
    FacetValueSelectorComponent,
    ItemsPerPageControlsComponent,
    PaginationControlsComponent,
    TableRowActionComponent,
    FacetValueChipComponent,
    FileSizePipe,
    FormFieldComponent,
    FormFieldControlDirective,
    FormItemComponent,
    ModalDialogComponent,
    PercentageSuffixInputComponent,
    DialogComponentOutletComponent,
    DialogButtonsDirective,
    DialogTitleDirective,
    SelectToggleComponent,
    LanguageSelectorComponent,
    RichTextEditorComponent,
    SimpleDialogComponent,
    TitleInputComponent,
    SentenceCasePipe,
    DropdownComponent,
    DropdownMenuComponent,
    SortPipe,
    DropdownTriggerDirective,
    DropdownItemDirective,
    OrderStateLabelComponent,
    FormattedAddressComponent,
    LabeledDataComponent,
    StringToColorPipe,
    ObjectTreeComponent,
    IfPermissionsDirective,
    IfMultichannelDirective,
    HasPermissionPipe,
    ActionBarItemsComponent,
    DisabledDirective,
    AssetFileInputComponent,
    AssetGalleryComponent,
    AssetPickerDialogComponent,
    EntityInfoComponent,
    DatetimePickerComponent,
    ChannelBadgeComponent,
    ChannelAssignmentControlComponent,
    ChannelLabelPipe,
    IfDefaultChannelActiveDirective,
    ExtensionHostComponent,
    CustomFieldLabelPipe,
];

@NgModule({
    imports: IMPORTS,
    exports: [...IMPORTS, ...DECLARATIONS],
    declarations: DECLARATIONS,
    providers: [
        // This needs to be shared, since lazy-loaded
        // modules have their own entryComponents which
        // are unknown to the CoreModule instance of ModalService.
        // See https://github.com/angular/angular/issues/14324#issuecomment-305650763
        ModalService,
        CanDeactivateDetailGuard,
    ],
    entryComponents: [ModalDialogComponent, SimpleDialogComponent, AssetPickerDialogComponent],
    schemas: [CUSTOM_ELEMENTS_SCHEMA],
})
export class SharedModule {}
