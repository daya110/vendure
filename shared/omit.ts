export type Omit<T, K> = Pick<T, Exclude<keyof T, K>>;

declare const File: any;

/**
 * Type-safe omit function - returns a new object which omits the specified keys.
 */
export function omit<T extends object, K extends keyof T>(obj: T, keysToOmit: K[]): Omit<T, K>;
export function omit<T extends object | any[], K extends keyof T>(obj: T, keysToOmit: string[], recursive: boolean): T;
export function omit<T extends any, K extends keyof T>(obj: T, keysToOmit: string[], recursive: boolean = false): T {
    if ((recursive && !isObject(obj)) || obj instanceof File) {
        return obj;
    }

    if (recursive && Array.isArray(obj)) {
        return obj.map((item: any) => omit(item, keysToOmit, true));
    }

    return Object.keys(obj).reduce((output: any, key) => {
        if (keysToOmit.includes(key)) {
            return output;
        }
        if (recursive) {
            return {...output, [key]: omit((obj as any)[key], keysToOmit, true)};
        }
        return {...output, [key]: (obj as any)[key]};
    }, {} as Omit<T, K>);
}

function isObject(input: any): input is object {
    return typeof input === 'object' && input !== null;
}
