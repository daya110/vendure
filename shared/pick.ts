/**
 * Returns a new object which is a subject of the input, including only the specified properties.
 * Can be called with a single argument (array of props to pick), in which case it returns a partially
 * applied pick function.
 */
export function pick<T extends string>(props: T[]): <U>(input: U) => { [K in T]: K extends keyof U ? U[K]: never };
export function pick<U, T extends keyof U>(input: U, props: T[]): { [K in T]: U[K] };
export function pick<T extends object, U extends T = any>(
    inputOrProps: U | Array<keyof T>,
    maybeProps?: Array<keyof T>,
): T | ((input: U) => T) {
    if (Array.isArray(inputOrProps)) {
        return (input: U) => _pick(input, inputOrProps);
    } else {
        return _pick(inputOrProps, maybeProps || []);
    }
}

function _pick<T extends object, U extends T = any>(input: U, props: Array<keyof T>): T {
    const output: any = {};
    for (const prop of props) {
        output[prop] = input[prop];
    }
    return output;
}
