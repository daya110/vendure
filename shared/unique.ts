/**
 * Returns an array with only unique values. Objects are compared by reference,
 * unless the `byKey` argument is supplied, in which case matching properties will
 * be used to check duplicates
 */
export function unique<T>(arr: T[], byKey?: keyof T): T[] {
    return arr.filter((item, index, self) => {
        return (
            index ===
            self.findIndex(i => {
                if (byKey === undefined) {
                    return i === item;
                } else {
                    return i[byKey] === item[byKey];
                }
            })
        );
    });
}
