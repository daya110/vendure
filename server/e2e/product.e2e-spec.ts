import {
    AddOptionGroupToProduct,
    AddOptionGroupToProductVariables,
    ApplyFacetValuesToProductVariants,
    ApplyFacetValuesToProductVariantsVariables,
    CreateProduct,
    CreateProduct_createProduct,
    CreateProductVariables,
    GenerateProductVariants,
    GenerateProductVariants_generateVariantsForProduct_variants,
    GenerateProductVariantsVariables,
    GetProductList,
    GetProductListVariables,
    GetProductWithVariants,
    GetProductWithVariantsVariables,
    LanguageCode,
    RemoveOptionGroupFromProduct,
    RemoveOptionGroupFromProductVariables,
    SortOrder,
    UpdateProduct,
    UpdateProductVariables,
    UpdateProductVariants,
    UpdateProductVariantsVariables,
} from 'shared/generated-types';
import { omit } from 'shared/omit';

import {
    ADD_OPTION_GROUP_TO_PRODUCT,
    APPLY_FACET_VALUE_TO_PRODUCT_VARIANTS,
    CREATE_PRODUCT,
    GENERATE_PRODUCT_VARIANTS,
    REMOVE_OPTION_GROUP_FROM_PRODUCT,
    UPDATE_PRODUCT,
    UPDATE_PRODUCT_VARIANTS,
} from '../../admin-ui/src/app/data/mutations/product-mutations';
import {
    GET_PRODUCT_LIST,
    GET_PRODUCT_WITH_VARIANTS,
} from '../../admin-ui/src/app/data/queries/product-queries';

import { TestClient } from './test-client';
import { TestServer } from './test-server';

describe('Product resolver', () => {
    const client = new TestClient();
    const server = new TestServer();

    beforeAll(async () => {
        const token = await server.init({
            logging: true,
            disableAuth: true,
            productCount: 20,
            customerCount: 1,
        });
        await client.init();
    }, 30000);

    afterAll(async () => {
        await server.destroy();
    });

    describe('products list query', () => {
        it('returns all products when no options passed', async () => {
            const result = await client.query<GetProductList, GetProductListVariables>(GET_PRODUCT_LIST, {
                languageCode: LanguageCode.en,
            });

            expect(result.products.items.length).toBe(20);
            expect(result.products.totalItems).toBe(20);
        });

        it('limits result set with skip & take', async () => {
            const result = await client.query<GetProductList, GetProductListVariables>(GET_PRODUCT_LIST, {
                languageCode: LanguageCode.en,
                options: {
                    skip: 0,
                    take: 3,
                },
            });

            expect(result.products.items.length).toBe(3);
            expect(result.products.totalItems).toBe(20);
        });

        it('filters by name', async () => {
            const result = await client.query<GetProductList, GetProductListVariables>(GET_PRODUCT_LIST, {
                languageCode: LanguageCode.en,
                options: {
                    filter: {
                        name: {
                            contains: 'fish',
                        },
                    },
                },
            });

            expect(result.products.items.length).toBe(2);
            expect(result.products.items[0].name).toBe('en Practical Frozen Fish');
            expect(result.products.items[1].name).toBe('en Sleek Wooden Fish');
        });

        it('sorts by name', async () => {
            const result = await client.query<GetProductList, GetProductListVariables>(GET_PRODUCT_LIST, {
                languageCode: LanguageCode.en,
                options: {
                    sort: {
                        name: SortOrder.ASC,
                    },
                },
            });

            expect(result.products.items.map(p => p.name)).toMatchSnapshot();
        });
    });

    describe('product query', () => {
        it('returns expected properties', async () => {
            const result = await client.query<GetProductWithVariants, GetProductWithVariantsVariables>(
                GET_PRODUCT_WITH_VARIANTS,
                {
                    languageCode: LanguageCode.en,
                    id: '2',
                },
            );

            if (!result.product) {
                fail('Product not found');
                return;
            }
            expect(omit(result.product, ['variants'])).toMatchSnapshot();
            expect(result.product.variants.length).toBe(2);
        });

        it('returns null when id not found', async () => {
            const result = await client.query<GetProductWithVariants, GetProductWithVariantsVariables>(
                GET_PRODUCT_WITH_VARIANTS,
                {
                    languageCode: LanguageCode.en,
                    id: 'bad_id',
                },
            );

            expect(result.product).toBeNull();
        });
    });

    describe('product mutation', () => {
        let newProduct: CreateProduct_createProduct;

        it('createProduct creates a new Product', async () => {
            const result = await client.query<CreateProduct, CreateProductVariables>(CREATE_PRODUCT, {
                input: {
                    image: 'baked-potato',
                    translations: [
                        {
                            languageCode: LanguageCode.en,
                            name: 'en Baked Potato',
                            slug: 'en-baked-potato',
                            description: 'A baked potato',
                        },
                        {
                            languageCode: LanguageCode.de,
                            name: 'de Baked Potato',
                            slug: 'de-baked-potato',
                            description: 'Eine baked Erdapfel',
                        },
                    ],
                },
            });
            newProduct = result.createProduct;
            expect(newProduct).toMatchSnapshot();
        });

        it('updateProduct updates a Product', async () => {
            const result = await client.query<UpdateProduct, UpdateProductVariables>(UPDATE_PRODUCT, {
                input: {
                    id: newProduct.id,
                    image: 'mashed-potato',
                    translations: [
                        {
                            languageCode: LanguageCode.en,
                            name: 'en Mashed Potato',
                            slug: 'en-mashed-potato',
                            description: 'A blob of mashed potato',
                        },
                        {
                            languageCode: LanguageCode.de,
                            name: 'de Mashed Potato',
                            slug: 'de-mashed-potato',
                            description: 'Eine blob von gemashed Erdapfel',
                        },
                    ],
                },
            });
            expect(result.updateProduct).toMatchSnapshot();
        });

        it('updateProduct errors with an invalid productId', async () => {
            try {
                await client.query<UpdateProduct, UpdateProductVariables>(UPDATE_PRODUCT, {
                    input: {
                        id: '999',
                        image: 'mashed-potato',
                        translations: [
                            {
                                languageCode: LanguageCode.en,
                                name: 'en Mashed Potato',
                                slug: 'en-mashed-potato',
                                description: 'A blob of mashed potato',
                            },
                            {
                                languageCode: LanguageCode.de,
                                name: 'de Mashed Potato',
                                slug: 'de-mashed-potato',
                                description: 'Eine blob von gemashed Erdapfel',
                            },
                        ],
                    },
                });
                fail('Should have thrown');
            } catch (err) {
                expect(err.message).toEqual(
                    expect.stringContaining("No Product with the id '999' could be found"),
                );
            }
        });

        it('addOptionGroupToProduct adds an option group', async () => {
            const result = await client.query<AddOptionGroupToProduct, AddOptionGroupToProductVariables>(
                ADD_OPTION_GROUP_TO_PRODUCT,
                {
                    optionGroupId: '1',
                    productId: newProduct.id,
                },
            );
            expect(result.addOptionGroupToProduct.optionGroups.length).toBe(1);
            expect(result.addOptionGroupToProduct.optionGroups[0].id).toBe('1');
        });

        it('addOptionGroupToProduct errors with an invalid productId', async () => {
            try {
                await client.query<AddOptionGroupToProduct, AddOptionGroupToProductVariables>(
                    ADD_OPTION_GROUP_TO_PRODUCT,
                    {
                        optionGroupId: '1',
                        productId: '999',
                    },
                );
                fail('Should have thrown');
            } catch (err) {
                expect(err.message).toEqual(
                    expect.stringContaining("No Product with the id '999' could be found"),
                );
            }
        });

        it('addOptionGroupToProduct errors with an invalid optionGroupId', async () => {
            try {
                await client.query<AddOptionGroupToProduct, AddOptionGroupToProductVariables>(
                    ADD_OPTION_GROUP_TO_PRODUCT,
                    {
                        optionGroupId: '999',
                        productId: newProduct.id,
                    },
                );
                fail('Should have thrown');
            } catch (err) {
                expect(err.message).toEqual(
                    expect.stringContaining("No OptionGroup with the id '999' could be found"),
                );
            }
        });

        it('removeOptionGroupFromProduct removes an option group', async () => {
            const result = await client.query<
                RemoveOptionGroupFromProduct,
                RemoveOptionGroupFromProductVariables
            >(REMOVE_OPTION_GROUP_FROM_PRODUCT, {
                optionGroupId: '1',
                productId: '1',
            });
            expect(result.removeOptionGroupFromProduct.optionGroups.length).toBe(0);
        });

        it('removeOptionGroupFromProduct errors with an invalid productId', async () => {
            try {
                await client.query<RemoveOptionGroupFromProduct, RemoveOptionGroupFromProductVariables>(
                    REMOVE_OPTION_GROUP_FROM_PRODUCT,
                    {
                        optionGroupId: '1',
                        productId: '999',
                    },
                );
                fail('Should have thrown');
            } catch (err) {
                expect(err.message).toEqual(
                    expect.stringContaining("No Product with the id '999' could be found"),
                );
            }
        });

        describe('variants', () => {
            let variants: GenerateProductVariants_generateVariantsForProduct_variants[];

            it('generateVariantsForProduct generates variants', async () => {
                const result = await client.query<GenerateProductVariants, GenerateProductVariantsVariables>(
                    GENERATE_PRODUCT_VARIANTS,
                    {
                        productId: newProduct.id,
                        defaultPrice: 123,
                        defaultSku: 'ABC',
                    },
                );
                variants = result.generateVariantsForProduct.variants;
                expect(variants.length).toBe(2);
                expect(variants[0].options.length).toBe(1);
                expect(variants[1].options.length).toBe(1);
            });

            it('generateVariantsForProduct throws with an invalid productId', async () => {
                try {
                    await client.query<GenerateProductVariants, GenerateProductVariantsVariables>(
                        GENERATE_PRODUCT_VARIANTS,
                        {
                            productId: '999',
                        },
                    );
                    fail('Should have thrown');
                } catch (err) {
                    expect(err.message).toEqual(
                        expect.stringContaining("No Product with the id '999' could be found"),
                    );
                }
            });

            it('updateProductVariants updates variants', async () => {
                const firstVariant = variants[0];
                const result = await client.query<UpdateProductVariants, UpdateProductVariantsVariables>(
                    UPDATE_PRODUCT_VARIANTS,
                    {
                        input: [
                            {
                                id: firstVariant.id,
                                translations: firstVariant.translations,
                                sku: 'ABC',
                                image: 'new-image',
                                price: 432,
                            },
                        ],
                    },
                );
                const updatedVariant = result.updateProductVariants[0];
                if (!updatedVariant) {
                    fail('no updated variant returned.');
                    return;
                }
                expect(updatedVariant.sku).toBe('ABC');
                expect(updatedVariant.image).toBe('new-image');
                expect(updatedVariant.price).toBe(432);
            });

            it('updateProductVariants throws with an invalid variant id', async () => {
                try {
                    await client.query<UpdateProductVariants, UpdateProductVariantsVariables>(
                        UPDATE_PRODUCT_VARIANTS,
                        {
                            input: [
                                {
                                    id: '999',
                                    translations: variants[0].translations,
                                    sku: 'ABC',
                                    image: 'new-image',
                                    price: 432,
                                },
                            ],
                        },
                    );
                    fail('Should have thrown');
                } catch (err) {
                    expect(err.message).toEqual(
                        expect.stringContaining("No ProductVariant with the id '999' could be found"),
                    );
                }
            });

            it('applyFacetValuesToProductVariants adds facets to variants', async () => {
                const result = await client.query<
                    ApplyFacetValuesToProductVariants,
                    ApplyFacetValuesToProductVariantsVariables
                >(APPLY_FACET_VALUE_TO_PRODUCT_VARIANTS, {
                    facetValueIds: ['1', '3', '5'],
                    productVariantIds: variants.map(v => v.id),
                });

                expect(result.applyFacetValuesToProductVariants.length).toBe(2);
                expect(result.applyFacetValuesToProductVariants[0].facetValues).toMatchSnapshot();
                expect(result.applyFacetValuesToProductVariants[1].facetValues).toMatchSnapshot();
            });

            it('applyFacetValuesToProductVariants errors with invalid facet value id', async () => {
                try {
                    await client.query<
                        ApplyFacetValuesToProductVariants,
                        ApplyFacetValuesToProductVariantsVariables
                    >(APPLY_FACET_VALUE_TO_PRODUCT_VARIANTS, {
                        facetValueIds: ['999', '888'],
                        productVariantIds: variants.map(v => v.id),
                    });
                    fail('Should have thrown');
                } catch (err) {
                    expect(err.message).toEqual(
                        expect.stringContaining("No FacetValue with the id '999' could be found"),
                    );
                }
            });

            it('applyFacetValuesToProductVariants errors with invalid variant id', async () => {
                try {
                    await client.query<
                        ApplyFacetValuesToProductVariants,
                        ApplyFacetValuesToProductVariantsVariables
                    >(APPLY_FACET_VALUE_TO_PRODUCT_VARIANTS, {
                        facetValueIds: ['1', '3', '5'],
                        productVariantIds: ['999'],
                    });
                    fail('Should have thrown');
                } catch (err) {
                    expect(err.message).toEqual(
                        expect.stringContaining("No ProductVariant with the id '999' could be found"),
                    );
                }
            });
        });
    });
});
