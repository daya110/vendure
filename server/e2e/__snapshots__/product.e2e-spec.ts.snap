// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product resolver product mutation createProduct creates a new Product 1`] = `
Object {
  "assets": Array [],
  "description": "A baked potato",
  "featuredAsset": null,
  "id": "21",
  "languageCode": "en",
  "name": "en Baked Potato",
  "optionGroups": Array [],
  "slug": "en-baked-potato",
  "translations": Array [
    Object {
      "description": "A baked potato",
      "languageCode": "en",
      "name": "en Baked Potato",
      "slug": "en-baked-potato",
    },
    Object {
      "description": "Eine baked Erdapfel",
      "languageCode": "de",
      "name": "de Baked Potato",
      "slug": "de-baked-potato",
    },
  ],
  "variants": Array [],
}
`;

exports[`Product resolver product mutation updateProduct updates a Product 1`] = `
Object {
  "assets": Array [],
  "description": "A blob of mashed potato",
  "featuredAsset": null,
  "id": "21",
  "languageCode": "en",
  "name": "en Mashed Potato",
  "optionGroups": Array [],
  "slug": "en-mashed-potato",
  "translations": Array [
    Object {
      "description": "A blob of mashed potato",
      "languageCode": "en",
      "name": "en Mashed Potato",
      "slug": "en-mashed-potato",
    },
    Object {
      "description": "Eine blob von gemashed Erdapfel",
      "languageCode": "de",
      "name": "de Mashed Potato",
      "slug": "de-mashed-potato",
    },
  ],
  "variants": Array [],
}
`;

exports[`Product resolver product mutation variants applyFacetValuesToProductVariants adds facets to variants 1`] = `
Array [
  Object {
    "code": "Harris_LLC",
    "id": "1",
    "name": "Harris LLC",
  },
  Object {
    "code": "Heathcote_-_Goyette",
    "id": "3",
    "name": "Heathcote - Goyette",
  },
  Object {
    "code": "Fisher,_Sporer_and_Bailey",
    "id": "5",
    "name": "Fisher, Sporer and Bailey",
  },
]
`;

exports[`Product resolver product mutation variants applyFacetValuesToProductVariants adds facets to variants 2`] = `
Array [
  Object {
    "code": "Harris_LLC",
    "id": "1",
    "name": "Harris LLC",
  },
  Object {
    "code": "Heathcote_-_Goyette",
    "id": "3",
    "name": "Heathcote - Goyette",
  },
  Object {
    "code": "Fisher,_Sporer_and_Bailey",
    "id": "5",
    "name": "Fisher, Sporer and Bailey",
  },
]
`;

exports[`Product resolver product query returns expected properties 1`] = `
Object {
  "assets": Array [
    Object {
      "fileSize": 4,
      "id": "1",
      "mimeType": "image/jpeg",
      "name": "charles-deluvio-695736-unsplash.jpg",
      "preview": "test-url/test-assets/charles-deluvio-695736-unsplash__preview.jpg",
      "source": "test-url/test-assets/charles-deluvio-695736-unsplash.jpg",
      "type": "IMAGE",
    },
    Object {
      "fileSize": 4,
      "id": "5",
      "mimeType": "image/jpeg",
      "name": "mikkel-bech-748940-unsplash.jpg",
      "preview": "test-url/test-assets/mikkel-bech-748940-unsplash__preview.jpg",
      "source": "test-url/test-assets/mikkel-bech-748940-unsplash.jpg",
      "type": "IMAGE",
    },
  ],
  "description": "en Accusantium sed libero repudiandae.",
  "featuredAsset": Object {
    "fileSize": 4,
    "id": "5",
    "mimeType": "image/jpeg",
    "name": "mikkel-bech-748940-unsplash.jpg",
    "preview": "test-url/test-assets/mikkel-bech-748940-unsplash__preview.jpg",
    "source": "test-url/test-assets/mikkel-bech-748940-unsplash.jpg",
    "type": "IMAGE",
  },
  "id": "2",
  "languageCode": "en",
  "name": "en Intelligent Cotton Salad",
  "optionGroups": Array [
    Object {
      "code": "size",
      "id": "1",
      "languageCode": "en",
      "name": "Size",
    },
  ],
  "slug": "en intelligent-cotton-salad",
  "translations": Array [
    Object {
      "description": "en Accusantium sed libero repudiandae.",
      "languageCode": "en",
      "name": "en Intelligent Cotton Salad",
      "slug": "en intelligent-cotton-salad",
    },
    Object {
      "description": "de Accusantium sed libero repudiandae.",
      "languageCode": "de",
      "name": "de Intelligent Cotton Salad",
      "slug": "de intelligent-cotton-salad",
    },
  ],
}
`;

exports[`Product resolver products list query sorts by name 1`] = `
Array [
  "en Awesome Cotton Keyboard",
  "en Awesome Granite Chair",
  "en Ergonomic Metal Sausages",
  "en Fantastic Granite Towels",
  "en Generic Frozen Tuna",
  "en Generic Metal Keyboard",
  "en Handcrafted Wooden Soap",
  "en Handmade Soft Hat",
  "en Handmade Wooden Car",
  "en Intelligent Cotton Salad",
  "en Intelligent Metal Shoes",
  "en Intelligent Rubber Chips",
  "en Intelligent Wooden Hat",
  "en Licensed Frozen Car",
  "en Practical Frozen Fish",
  "en Practical Rubber Pants",
  "en Rustic Steel Gloves",
  "en Small Wooden Chair",
  "en Tasty Soft Gloves",
  "en Unbranded Fresh Hat",
]
`;
